java-client/session/src/main/java/io/deephaven/client/impl/SessionImpl.java

```diff
@@ -0,0 +1,342 @@
+import io.deephaven.proto.backplane.script.grpc.BindTableToVariableResponse;
+import io.deephaven.proto.backplane.script.grpc.ConsoleServiceGrpc.ConsoleServiceStub;
+import io.grpc.CallCredentials;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.lang.model.SourceVersion;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A {@link Session} implementation that uses
+ * {@link io.deephaven.proto.backplane.grpc.BatchTableRequest batch requests} and memoizes based on
+ * {@link io.deephaven.qst.table.TableSpec} equality.
+ *
+ * <p>
+ * {@inheritDoc}
+ */
+public final class SessionImpl extends SessionBase {
+
+    private static final int REFRESH_RETRIES = 5;
+
+    public interface Handler {
+        void onRefreshSuccess();
+
+        void onRefreshTokenError(Throwable t, Runnable invokeForRetry);
+
+        void onCloseSessionError(Throwable t);
+
+        void onClosed();
+    }
+
+    private static class Retrying implements Handler {
+        private static final Logger log = LoggerFactory.getLogger(Retrying.class);
+
+        private final int maxRefreshes;
+        private int remainingRefreshes;
+
+        Retrying(int maxRefreshes) {
+            this.maxRefreshes = maxRefreshes;
+        }
+
+        @Override
+        public void onRefreshSuccess() {
+            remainingRefreshes = maxRefreshes;
+        }
+
+        @Override
+        public void onRefreshTokenError(Throwable t, Runnable invokeForRetry) {
+            if (remainingRefreshes > 0) {
+                remainingRefreshes--;
+                log.warn("Error refreshing token, trying again", t);
+                invokeForRetry.run();
+                return;
+            }
+            log.error("Error refreshing token, giving up", t);
+        }
+
+        @Override
+        public void onCloseSessionError(Throwable t) {
+            log.error("onCloseSessionError", t);
+        }
+
+        @Override
+        public void onClosed() {
+
+        }
+    }
+
+    public static SessionImpl create(SessionServiceGrpc.SessionServiceBlockingStub stubBlocking,
+        SessionServiceGrpc.SessionServiceStub stub, TableServiceStub tableServiceStub,
+        ConsoleServiceStub consoleServiceStub, ScheduledExecutorService executor) {
+        HandshakeRequest request = initialHandshake();
+        HandshakeResponse response = stubBlocking.newSession(request);
+        AuthenticationInfo initialAuth = AuthenticationInfo.of(response);
+        SessionImpl session = new SessionImpl(executor, stub, tableServiceStub, consoleServiceStub,
+            new Retrying(REFRESH_RETRIES), initialAuth);
```

Comment:
It looks like these session state hooks in Retrying (new session, and giving up on refresh) might be nice events for a client to hook into. For example, can I build a client that maintains a subscription to a table on a worker over a flaky network connection or to a worker that restarts frequently?